src/compiler.ts

```diff
@@ -6512,14 +6598,64 @@ export class Compiler extends DiagnosticEmitter {
         return module.unreachable();
       }
     }
-    return this.compileCallIndirect(
-      assert(signature), // FIXME: asc can't see this yet
-      indexArg,
-      expression.arguments,
-      expression,
-      0,
-      contextualType == Type.void
-    );
+
+    var returnType = signature.returnType;
+    var tempResultLocal = this.currentFlow.getTempLocal(returnType);
+    var tempFunctionReferenceLocal = this.currentFlow.getTempLocal(returnType);
+
+    var callExpr = module.block(null, [
+      module.local_set(
+        tempFunctionReferenceLocal.index,
+        indexArg
+      ),
+      module.push(
+        module.global_get(BuiltinNames.global_closure, NativeType.I32)
+      ),
+      this.compileCallIndirect(
+        assert(signature), // FIXME: asc can't see this yet
+        module.if(
+          module.binary(
+            BinaryOp.EqI32,
+            module.binary(
+              BinaryOp.AndI32,
+              module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+              module.i32(1 << 31)
+            ),
+            module.i32(1 << 31)
+          ),
```

Comment:
This is checking for our special closure pointer which is in this binary format:
[1000][all but the last 4 bits of the context address]

A normal function would never have that MSB set because we instituted a limit below 1 << 31.